Use GoStringer interface for structs that support it

ruzz311 opened the pull request:
This just changes the great work @otaku has already done to use GoStringer since it is meant for fmt.printf-type statements.  I don't think an additional "(stringer)" message is needed in the output because if someone implements a GoStringer, they've made the choice to see the resulting string in print statements.

kr commented on formatter_test.go:
```diff
@@ -259,3 +260,46 @@ func TestCycle(t *testing.T) {
+func Test(t *testing.T) {
+	date := time.Date(2006, 1, 2, 3, 4, 5, .678901e9, time.Local)
+
+	s := fmt.Sprintf("%s", Formatter(date))
```
Formatter ignores format verbs other than v with # and ' ' (space) flags. This format string needs to use `% #v` instead of `%s`.

ruzz311 replied:
I guess I forgot to review the tests before PR - I'll update them and push.  Thanks for the quick review.

kr commented on formatter_test.go:
```diff
@@ -259,3 +260,46 @@ func TestCycle(t *testing.T) {
+func TestStringer(t *testing.T) {
+	s := new(StringerStruct)
+
+	str := fmt.Sprintf("%s", Formatter(s))
```
Formatter ignores format verbs other than v with # and ' ' (space) flags. This format string needs to use `% #v` instead of `%s`. Currently, it's not testing what it's meant to test.

kr commented on formatter_test.go:
```diff
@@ -259,3 +260,46 @@ func TestCycle(t *testing.T) {
+	s := new(MarshalTextStruct)
+
+	m, _ := s.MarshalText()
+	str := fmt.Sprintf("%s", Formatter(m))
```
Formatter ignores format verbs other than v with # and ' ' (space) flags. This format string needs to use `% #v` instead of `%s`. Currently, it's not testing what it's meant to test.

ruzz311 commented:
@kr - I removed the TextMarshaler / Stringer in favor of GoStringer.  Also, fixed tests with a test using a date example similar to the code I was working on that inspired this PR (had to make a custom date type for an UnmarshalXML).

Also, I was looking at a couple other tests and noticed diff_test.go did not cover maps at all so I added a few test cases and raised the code coverage for that file almost 30% (nearly 10% for the whole package)
